File: day_02/p2.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>

using ints = std::vector<int>;

std::vector<ints> ParseInput(const std::string& fname)
{
    std::vector<ints> records {};
    const auto lines = pypp::splitFileLines(fname);

    for (const auto& line : lines) {
        ints record {};
        auto record_str = pypp::split(line, ' ');

        std::ranges::transform(record_str,
            std::back_inserter(record),
            [](const std::string& s) { return std::stoi(s); });

        records.push_back(record);
    }

    return records;
}

bool IsSafeByOrderAndDifference(const ints& record)
{
    ints diffs {};

    for (const auto& tuple : pypp::zip(record.begin() + 1, record.end(), record.begin()))
        diffs.push_back(std::get<1>(tuple) - std::get<0>(tuple));

    return std::ranges::all_of(diffs, [](const int diff) { return 1 <= diff && diff <= 3; })
        || std::ranges::all_of(diffs, [](const int diff) { return -3 <= diff && diff <= -1; });
}

std::vector<ints> generate_mutations(const ints& record)
{
    std::vector<ints> mutations;
    for (std::size_t i = 0; i < record.size(); ++i) {
        ints mutation;
        std::ranges::copy_if(record, std::back_inserter(mutation), [i, j = 0](int) mutable { return j++ != i; });
        mutations.push_back(std::move(mutation));
    }
    mutations.push_back(record); // original
    return mutations;
}

int Compute(const std::vector<ints>& records)
{
    int sum { 0 };

    for (const auto& record : records)
        if (std::ranges::any_of(generate_mutations(record), [](const ints& list) { return IsSafeByOrderAndDifference(list); }))
            sum++;

    return sum;
}

int main()
{
    auto records = ParseInput("input.txt");
    std::cout << Compute(records) << std::endl;

    return 0;
}

File: day_02/p2.py
from typing import Generator, List


def parse_input(fname: str) -> List[List[int]]:
    lines = []
    with open(fname, 'r') as f:
        lines = f.read().splitlines()
    
    return [list(map(int, line.split())) for line in lines]

def is_safe_by_order_and_difference(record: List[int]) -> bool:
    
    diffs = [a-b for (a, b) in zip(record, record[1:])]

    return all(1 <= diff <= 3 for diff in diffs) or \
           all(-3 <= diff <= -1 for diff in diffs)

def generate_mutations(record: List[int]) -> Generator[List[int]]:
    """
    Generate sub-lists of the given list, excluding one element at a time,
    followed by the whole list.
    
    Args:
       record : The list to generate sub-lists from.
    
    Yields:
       A sublist of the original list with one element excluded at a time,
       and finally the original list.
       
    Example:
        [1, 2, 3, 4, 5] generates:
        - [2, 3, 4, 5]
        - [1, 3, 4, 5]
        - [1, 2, 4, 5]
        - [1, 2, 3, 5]
        - [1, 2, 3, 4, 5]], i.e. the original list
    """
    for i in range(len(record)):
        yield record[:i] + record[i+1:]
    
    yield record # original

def compute(records: List[List[int]]) -> int:
    return sum(
        any(is_safe_by_order_and_difference(modified)
            for modified in generate_mutations(record))  
        for record in records)

def main():
    records = parse_input("input.txt")
    print(compute(records))

if __name__ == "__main__":
    main()
